File: Practice/20/C++/task20/task20.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace std;

int main()
{
	setlocale(LC_ALL, "Russian");
	vector<string> name;
	vector<int> price;
	vector<int> volume;
	vector<int> purch;
	vector<int> order;
	map<int, double> d;
	string na;
	int k, money, money0, pr, vo, m, j, a = 0, v = 0;
	bool f = true;
	cout << "Деньги: ";
	cin >> money;
	money0 = money;
	cout << "Количество видов напитков в магазине: ";
	cin >> k;
	for (int i = 0; i < k; i++) {   // цикл ввода данных и рассчёта цены за литр
		cout << i + 1 << " напиток: ";
		cin >> na >> pr >> vo;
		name.push_back(na);
		price.push_back(pr);
		volume.push_back(vo);
		purch.push_back(0);
		d[i] = pr / vo;
	}
	while (size(d) > 0) {   // цикл составления списка порядка покупок
		m = 1000000;
		for (auto i: d) {
			if (i.second < m) {
				m = i.second;
				j = i.first;
			}
		}
		order.push_back(j);
		d.erase(j);
	}
	while (f) {   // цикл составления списка покупок
		if (money >= price[order[a]]) {
			money -= price[order[a]];
			purch[order[a]] += 1;
		}
		else {
			if (a < k - 1) {
				a += 1;
			}
			else {
				f = false;
			}
		}
	}
	if (money < money0) {
		cout << "Надо купить:\n";
		for (int i = 0; i < k; i++) {   // цикл вывода покупок и расчёта суммарного объёма
			v = v + purch[i] * volume[i];
			if (purch[i] > 0) {
				cout << name[i] << ": " << purch[i] << " шт.\n";
			}
		}
		cout << "Суммарный объём: " << v << endl;
		cout << "Осталось денег: " << money;
	}
	else {
		cout << -1;
	}
	return 0;
}

File: Practice/20/Python/task20.py
money = int(input("деньги: "))
money0 = money
k = int(input("количество видов напитков в магазине: "))
purchase = {}   # словарь показывающий количество купленных напитков
order = []   # список ключей, показывающий в каком порядке надо будет покупать напитки
drinks = {}   # словарь напитков: 0: "назв", цена, объём
d = {}   # цена за литр для каждого напитка
for i in range(k):
	print(str(i + 1) + ": ", end = "")
	name, price, volume = input().split(" ")
	price = int(price)
	volume = int(volume)
	drinks[i] = (name, price, volume)
	d[i] = price / volume
	purchase[i] = 0
while (len(d) > 0):   # цикл составления списка order
	m = 1000000
	for i in d.keys():
		if (m > d[i]):
			m = d[i]
			j = i
	order.append(j)
	del d[j]
a = 0
f = True
while f:
	if (money >= drinks[order[a]][1]):
		money -= drinks[order[a]][1]
		purchase[order[a]] += 1
	elif (a < k-1):
		a += 1
	else:
		f = False
if (money < money0):
	print("надо купить:")
	for i in range(k):
		if (purchase[i] != 0):
			print(str(drinks[i][0]) + ": " + str(purchase[i]) + " шт.")
	v = 0
	for i in range(k):
		v = v + purchase[i] * drinks[i][2]
	print("суммарный объём: " + str(v))
	print("осталось денег: " + str(money))
else:
	print(-1)
input("Нажмите Enter")
